File: src/services/FileSystemService.cpp
// src/services/FileSystemService.cpp

#include "../../include/services/FileSystemService.h"
#include "../../include/storage/Storage.h"
#include "../../include/services/FileService.h"
#include "../../include/services/FolderService.h"
#include "../../include/services/HistoryService.h"
#include "../../include/services/GrepService.h"
#include <vector>
#include <string>
#include <map>
#include <iostream>
#include <stack>

using namespace std;

void FileSystemService::createFile(string folderId, string fileName) 
{ 
    fileService->createFile(folderId, fileName); 
    historyService->addEntry("touch " + fileName, "CREATE_FILE", fileName, currentPath());
}

string FileSystemService::getCurrentFolder() { return folderService->getCurrentFolder(); }

void FileSystemService::addContent(string fileId, string content) 
{ 
    fileService->addContent(fileId, content); 
    historyService->addEntry("write " + fileId + " " + content, "WRITE_FILE", fileId, currentPath());
}

void FileSystemService::removeFile(string fileName) 
{ 
    fileService->removeFile(fileName); 
    historyService->addEntry("rm " + fileName, "REMOVE_FILE", fileName, currentPath());
}

string FileSystemService::showFileContent(string fileId) { return fileService->showFileContent(fileId); }

void FileSystemService::createFolder(string parentFolderId, string folderName) 
{ 
    folderService->createFolder(parentFolderId, folderName); 
    historyService->addEntry("mkdir " + folderName, "CREATE_FOLDER", folderName, currentPath());
}

void FileSystemService::removeFolder(string folderName) 
{ 
    folderService->removeFolder(folderName); 
    historyService->addEntry("rmdir " + folderName, "REMOVE_FOLDER", folderName, currentPath());
}

void FileSystemService::showTree(string folderId) 
{ 
    folderService->showTree(folderService->getCurrentFolder()); 
    historyService->addEntry("tree", "SHOW_TREE", "", currentPath());
}

void FileSystemService::listAllItems(string folderId) 
{ 
    folderService->listAllItems(folderId); 
    historyService->addEntry("ls", "LIST_ITEMS", "", currentPath());
}

void FileSystemService::getIntoFolder(string folderName) 
{ 
    folderService->getIntoFolder(folderName); 
    historyService->addEntry("cd " + folderName, "CHANGE_DIR", folderName, currentPath());
}

bool FileSystemService::isFolderAvailable(string name) { return Storage::getInstance()->validateFolder(name); }

string FileSystemService::currentPath() { return Storage::getInstance()->getPath(folderService->getCurrentFolder()); }

// History operations
void FileSystemService::showHistory() const
{
    historyService->showHistory();
}

void FileSystemService::showHistory(int count) const
{
    historyService->showHistory(count);
}

void FileSystemService::clearHistory()
{
    historyService->clearHistory();
}

int FileSystemService::getHistoryCount() const
{
    return historyService->getHistoryCount();
}

// Grep operations
void FileSystemService::grepPattern(const string& pattern)
{
    grepService->grep(pattern);
    historyService->addEntry("grep " + pattern, "GREP", pattern, currentPath());
}

void FileSystemService::grepInFile(const string& pattern, const string& fileName)
{
    grepService->grepInFile(pattern, fileName);
    historyService->addEntry("grep " + pattern + " " + fileName, "GREP_FILE", fileName, currentPath());
}

void FileSystemService::grepRecursive(const string& pattern)
{
    GrepOptions options;
    options.recursive = true;
    grepService->grep(pattern, options);
    historyService->addEntry("grep -r " + pattern, "GREP_RECURSIVE", pattern, currentPath());
}

void FileSystemService::grepWithOptions(const string& pattern, const string& options)
{
    GrepOptions grepOpts;
    
    // Parse options
    for (char c : options) {
        switch (c) {
            case 'i': grepOpts.caseInsensitive = true; break;
            case 'r': grepOpts.recursive = true; break;
            case 'c': grepOpts.countOnly = true; break;
            case 'v': grepOpts.invertMatch = true; break;
            case 'n': grepOpts.showLineNumbers = true; break;
        }
    }
    
    grepService->grep(pattern, grepOpts);
    historyService->addEntry("grep -" + options + " " + pattern, "GREP_OPTIONS", pattern, currentPath());
}

void FileSystemService::showGrepHelp()
{
    grepService->showGrepHelp();
    historyService->addEntry("grep --help", "GREP_HELP", "", currentPath());
}

FileSystemService::FileSystemService()
{
    folderService = new FolderService();
    fileService = new FileService();
    historyService = new HistoryService();
    grepService = new GrepService();
}

File: src/services/FileSystemService.js
const Storage = require('../storage/Storage');
const FileService = require('./FileService');
const FolderService = require('./FolderService');

/**
 * FileSystemService integrates file and folder services
 * Migrated from C++ FileSystemService.h/FileSystemService.cpp
 */
class FileSystemService {
  constructor() {
    this.store = Storage.getInstance();
    this.fileService = new FileService();
    this.folderService = new FolderService();
  }

  /**
   * Create a file in the specified folder
   * @param {string} folderId - Folder ID where the file will be created
   * @param {string} fileName - Name of the file
   * @returns {Object} - Created file or null if creation failed
   */
  createFile(folderId, fileName) {
    return this.fileService.createFile(folderId, fileName);
  }

  /**
   * Get current folder ID
   * @returns {string} - Current folder ID
   */
  getCurrentFolder() {
    return this.folderService.getCurrentFolder();
  }

  /**
   * Add content to a file
   * @param {string} fileName - Name of the file
   * @param {string} content - Content to add
   */
  addContent(fileName, content) {
    this.fileService.addContent(fileName, content);
  }

  /**
   * Remove a file by name
   * @param {string} fileName - Name of the file to remove
   * @returns {boolean} - True if removal was successful
   */
  removeFile(fileName) {
    return this.fileService.removeFile(fileName);
  }

  /**
   * Get file content
   * @param {string} fileId - ID of the file
   * @returns {string} - Content of the file
   */
  showFileContent(fileId) {
    return this.fileService.showFileContent(fileId);
  }

  /**
   * Create a folder in the specified parent folder
   * @param {string} parentFolderId - Parent folder ID
   * @param {string} folderName - Name of the folder to create
   * @returns {Object} - Created folder or null if creation failed
   */
  createFolder(parentFolderId, folderName) {
    return this.folderService.createFolder(parentFolderId, folderName);
  }

  /**
   * Remove a folder by name
   * @param {string} folderName - Name of the folder to remove
   * @returns {boolean} - True if removal was successful
   */
  removeFolder(folderName) {
    return this.folderService.removeFolder(folderName);
  }

  /**
   * Get folder tree starting from specified folder
   * @param {string} folderId - Starting folder ID
   * @returns {Object} - Tree structure
   */
  showTree(folderId) {
    return this.folderService.showTree(folderId);
  }

  /**
   * List all items in a folder
   * @param {string} folderId - Folder ID
   * @returns {Array} - Array of items in the folder
   */
  listAllItems(folderId) {
    return this.folderService.listAllItems(folderId);
  }

  /**
   * Navigate into a folder
   * @param {string} folderName - Folder name
   * @returns {boolean} - True if navigation was successful
   */
  getIntoFolder(folderName) {
    return this.folderService.getIntoFolder(folderName);
  }

  /**
   * Check if a folder exists in current directory
   * @param {string} name - Name of the folder
   * @returns {boolean} - True if folder exists
   */
  isFolderAvailable(name) {
    return this.folderService.validateFolder(name);
  }

  /**
   * Get current path as a string
   * @returns {string} - Current path
   */
  currentPath() {
    return this.store.getPath(this.folderService.getCurrentFolder());
  }
  
  /**
   * Get file by name
   * @param {string} fileName - Name of the file
   * @returns {Object|null} - File details or null if not found
   */
  getFileByName(fileName) {
    return this.fileService.getFileByName(fileName);
  }
  
  /**
   * Get folder by name
   * @param {string} folderName - Name of the folder
   * @returns {Object|null} - Folder details or null if not found
   */
  getFolderByName(folderName) {
    return this.folderService.getFolderByName(folderName);
  }
}

module.exports = FileSystemService;